ykrt/src/compile/j2/opt/known_bits.rs

```diff
@@ -0,0 +1,398 @@
+                    kind: ConstKind::Int(res.as_arbbitint()),
+                }));
+            }
+
+            // The `and` operation adds new information in the form of set zero. E.g. `unknown &
+            // (~1)` zeroes the least significant bit. If the result has no new set zeroes,
+            // that means this op is useless.
+            if rhs_b.all_known()
+                && rhs_b
+                    .zeroes()
+                    .bitand(&lhs_b.known_ones().bitor(&lhs_b.unknowns))
+                    .count_ones()
+                    == 0
+            {
+                return OptOutcome::Equiv(lhs);
+            }
+        }
+        OptOutcome::Rewritten(inst.into())
+    }
+
+    fn opt_const(&mut self, inst: Const) -> OptOutcome {
+        let Const { tyidx: _, kind } = &inst;
+        if let ConstKind::Int(kind) = kind {
+            self.set_pending(KnownBitValue::from_const(kind.clone()))
+        }
+        OptOutcome::Rewritten(inst.into())
+    }
+
+    // The `or` operation adds new information in the form of set ones.  E.g. `unknown | (1)` sets
+    // the least significant bit to one. If the result has no new set ones, that means this op is
+    // useless.
+    fn opt_or(&mut self, opt: &PassOpt, mut inst: Or) -> OptOutcome {
+        inst.canonicalise(opt);
+        let Or {
+            tyidx,
+            lhs,
+            rhs,
+            disjoint: _,
+        } = inst;
+        if let Some(lhs_b) = self.as_knownbits(opt, lhs)
+            && let Some(rhs_b) = self.as_knownbits(opt, rhs)
+        {
+            let res = lhs_b.bitor(&rhs_b);
+            self.set_pending(res.clone());
+
+            // If we know the output's bits, emit that.
+            if res.all_known() {
+                return OptOutcome::Rewritten(Inst::Const(Const {
+                    tyidx,
+                    kind: ConstKind::Int(res.as_arbbitint()),
+                }));
+            }
+
+            // If no new information (set ones) was gained, that means this
+            // op is useless.
+            if rhs_b.all_known()
+                && rhs_b
+                    .known_ones()
+                    .bitand(&lhs_b.zeroes().bitor(&lhs_b.unknowns))
+                    .count_ones()
+                    == 0
+            {
+                return OptOutcome::Equiv(lhs);
+            }
+        }
+        OptOutcome::Rewritten(inst.into())
+    }
+}
+
+/// Known bits for a single value.
+///
+/// In short:
+/// | one | unknown | knownbit |
+/// |-----|---------|----------|
+/// | 0   | 1       | ?        |
+/// | 0   | 0       | 0        |
+/// | 1   | 0       | 1        |
+/// | 1   | 1       | illegal  |
+///
+/// To ensure monotonicity,transitions from `?` to` 0` or `1` are valid, but not the other way
+/// around. `illegal` occurs when both `0` and `1` are set and known, which is impossible in a
+/// valid program. `illegal` indicates a likely bug in the optimizer/IR.
+#[derive(Clone)]
+pub struct KnownBitValue {
```

Comment:
Oops, just realised: why are these `pub`? They shouldn't be known outside this file AFAICS?